File: scripts/case_selection/export_bpc_selected_cases.py
#!/usr/bin/env python
# Authors: Xindi Guo, Kristen Dang
# September 17, 2019

import argparse
import pandas as pd
import synapseclient
import os
from synapseclient import Synapse
from synapseclient import Activity, File
import numpy as np
from datetime import date

# user input --------------------------
parser = argparse.ArgumentParser()
parser.add_argument(
    "-i",
    "--input",
    help="Synapse ID of the input file that has the BPC selected cases",
    required=True,
)
parser.add_argument(
    "-o",
    "--output",
    default="syn20798271",
    help="Synapse ID of the BPC output folder. Default: syn20798271",
)
parser.add_argument(
    "-p",
    "--phase",
    help="BPC phase. i.e. pilot, phase 1, phase 1 additional",
    required=True,
)
parser.add_argument(
    "-c", "--cohort", help="BPC cohort. i.e. NSCLC, CRC, BrCa, and etc.", required=True
)
parser.add_argument(
    "-s", "--site", help="BPC site. i.e. DFCI, MSK, UHN, VICC, and etc.", required=True
)
args = parser.parse_args()

in_file = args.input
out_folder = args.output
phase = args.phase
cohort = args.cohort
site = args.site

# check user input -----------------
# TODO why are the inputs different...
phase_option = ["phase 1", "phase 1 additional", "phase 2"]
cohort_option = ["NSCLC", "CRC", "BrCa", "PANC", "Prostate", "BLADDER"]
site_option = ["DFCI", "MSK", "UHN", "VICC"]

assert (
    phase in phase_option
), f"Error: {phase} is not a valid phase. Valid values: {', '.join(phase_option)}"
assert (
    cohort in cohort_option
), f"Error: {cohort} is not a valid cohort. Valid values: {', '.join(cohort_option)}"
assert (
    site in site_option
), f"Error: {site} is not a valid site. Valid values: {', '.join(site_option)}"

# setup --------------------------
syn = Synapse()
syn.login()

# clinical data
# always use the most recent consortium release - Feb 2022
clinical_sample_id = "syn9734573"
clinical_patient_id = "syn9734568"
# HACK pin version
clinical_sample_id = "syn51499964"
clinical_patient_id = "syn51499962"

# mapping tables
sex_mapping = syn.tableQuery("SELECT * FROM syn7434222").asDataFrame()
race_mapping = syn.tableQuery("SELECT * FROM syn7434236").asDataFrame()
ethnicity_mapping = syn.tableQuery("SELECT * FROM syn7434242").asDataFrame()

# output setup
phase_no_space = phase.replace(" ", "_")
output_entity_name = f"{site}_{cohort}_{phase_no_space}_genie_export.csv"
output_file_name = f"{site}_{cohort}_{phase_no_space}_genie_export_{date.today()}.csv"

# download input file and get selected cases/samples
selected_info = pd.read_csv(syn.get(in_file).path)
selected_cases = selected_info["PATIENT_ID"]
selected_samples = ";".join(selected_info["SAMPLE_IDS"]).split(";")

# create the data file ----------------------------

# Create query for selected cases
temp = ", ".join([f"'{item}'" for item in selected_samples])

# download clinical data
# sample clinical data
clinical_sample = pd.read_csv(
    syn.get(clinical_sample_id, followLink=True).path,
    comment="#",
    sep="\t",
)
clinical_sample = clinical_sample[clinical_sample["SAMPLE_ID"].isin(selected_samples)]

# patient clinical data
clinical_patient = pd.read_csv(
    syn.get(clinical_patient_id, followLink=True).path,
    comment="#",
    sep="\t",
)

# combined clinical data
clinical = pd.merge(clinical_sample, clinical_patient, on="PATIENT_ID", how="left")

# change the columns to lower case
clinical.columns = map(str.lower, clinical.columns)

# Get all samples for those patients
samples_per_patient = [
    clinical["sample_id"][clinical["patient_id"] == x].tolist() for x in selected_cases
]


def remap_clinical_values(
    clinicaldf: pd.DataFrame,
    sex_mapping: pd.DataFrame,
    race_mapping: pd.DataFrame,
    ethnicity_mapping: pd.DataFrame,
) -> pd.DataFrame:
    """Remap clinical attributes from integer to string values

    Args:
        clinicaldf: Clinical data
        sex_mapping: Sex mapping data
        race_mapping: Race mapping data
        ethnicity_mapping: Ethnicity mapping data

    Returns:
        Mapped clinical dataframe
    """

    race_mapping.index = race_mapping["CBIO_LABEL"]
    race_dict = race_mapping.to_dict()

    ethnicity_mapping.index = ethnicity_mapping["CBIO_LABEL"]
    ethnicity_dict = ethnicity_mapping.to_dict()

    sex_mapping.index = sex_mapping["CBIO_LABEL"]
    sex_dict = sex_mapping.to_dict()

    # Use pandas mapping feature
    clinicaldf = clinicaldf.replace(
        {
            "PRIMARY_RACE": race_dict["CODE"],
            "SECONDARY_RACE": race_dict["CODE"],
            "TERTIARY_RACE": race_dict["CODE"],
            "SEX": sex_dict["CODE"],
            "ETHNICITY": ethnicity_dict["CODE"],
        }
    )

    return clinicaldf


subset_patient = clinical_patient[clinical_patient["PATIENT_ID"].isin(selected_cases)]
# TODO: these mappings go from non-granular to granular CODES
# TODO which _could be_ an issue.  Check about these mappings..
subset_patient = remap_clinical_values(
    clinicaldf=subset_patient,
    sex_mapping=sex_mapping,
    race_mapping=race_mapping,
    ethnicity_mapping=ethnicity_mapping,
)
subset_patient.columns = map(str.lower, subset_patient.columns)

# mapping data for each instrument
# instrument - patient_characteristics
patient_output = pd.DataFrame({"record_id": selected_cases})
patient_output["redcap_repeat_instrument"] = ""
patient_output["redcap_repeat_instance"] = ""

patient_output["genie_patient_id"] = patient_output["record_id"]
patient_output.set_index("genie_patient_id", inplace=True, drop=False)
patient_output.loc[subset_patient["patient_id"], "birth_year"] = subset_patient[
    "birth_year"
].to_list()
patient_output.loc[
    subset_patient["patient_id"], "naaccr_ethnicity_code"
] = subset_patient["ethnicity"].to_list()
patient_output.loc[
    subset_patient["patient_id"], "naaccr_race_code_primary"
] = subset_patient["primary_race"].to_list()
patient_output.loc[
    subset_patient["patient_id"], "naaccr_race_code_secondary"
] = subset_patient["secondary_race"].to_list()
patient_output.loc[
    subset_patient["patient_id"], "naaccr_race_code_tertiary"
] = subset_patient["tertiary_race"].to_list()
patient_output.loc[subset_patient["patient_id"], "naaccr_sex_code"] = subset_patient[
    "sex"
].to_list()

# recode
# cannotReleaseHIPAA = NA
patient_output["birth_year"].replace("cannotReleaseHIPAA", np.nan, inplace=True)
# -1 Not collected = 9 Unknown
patient_output["naaccr_ethnicity_code"].replace(-1, 9, inplace=True)
# -1 Not collected = 99 Unknown
patient_output["naaccr_race_code_primary"].replace(-1, 99, inplace=True)
# -1 Not collected = 88 according to NAACCR
patient_output["naaccr_race_code_secondary"].replace(-1, 88, inplace=True)
patient_output["naaccr_race_code_tertiary"].replace(-1, 88, inplace=True)

# instrument - cancer_panel_test
sample_info_list = []
for samples in samples_per_patient:
    sample_list = []
    for i, sample in enumerate(samples, start=1):
        temp_df = pd.DataFrame(
            {"record_id": clinical["patient_id"][clinical["sample_id"] == sample]}
        )
        temp_df["redcap_repeat_instrument"] = "cancer_panel_test"
        temp_df["redcap_repeat_instance"] = i
        temp_df["redcap_data_access_group"] = clinical["center"][
            clinical["sample_id"] == sample
        ]

        temp_df["cpt_genie_sample_id"] = sample
        temp_df["cpt_oncotree_code"] = clinical["oncotree_code"][
            clinical["sample_id"] == sample
        ]
        temp_df["cpt_sample_type"] = clinical["sample_type_detailed"][
            clinical["sample_id"] == sample
        ]
        temp_df["cpt_seq_assay_id"] = clinical["seq_assay_id"][
            clinical["sample_id"] == sample
        ]
        temp_df["cpt_seq_date"] = clinical["seq_year"][clinical["sample_id"] == sample]
        temp_df["age_at_seq_report"] = clinical["age_at_seq_report_days"][
            clinical["sample_id"] == sample
        ]
        sample_list.append(temp_df)
    combined_df = pd.concat(sample_list)
    sample_info_list.append(combined_df)

sample_info_df = pd.concat(sample_info_list)
patient_output = pd.concat([patient_output, sample_info_df])

# output and upload ----------------------------
from genie import process_functions


with open(output_file_name, "w") as out_f:
    no_float_text = process_functions.removePandasDfFloat(patient_output)
    convert_to_csv = no_float_text.replace("\t", ",")
    out_f.write(convert_to_csv)
# patient_output.to_csv(output_file_name, index=False, na_rep="")

# create an Activity
act = Activity(
    name="export main GENIE data",
    description="Export selected BPC patient data from main GENIE database",
    used=[clinical_sample_id, clinical_patient_id, in_file],
    executed="https://github.com/Sage-Bionetworks/genie-bpc-pipeline/tree/develop/scripts/case_selection/export_bpc_selected_cases.R",
)

# create a File
syn_file = File(
    path=output_file_name,
    parent=out_folder,
    name=output_entity_name,
    annotations={"phase": phase, "cohort": cohort, "site": site},
)

# TODO add this back in
# # store the File in Synapse
# syn_file = syn.store(syn_file)

# # set the provenance for the File
# syn.setProvenance(syn_file, act)

# remove the local file
# os.remove(output_file_name)

File: scripts/case_selection/export_bpc_selected_cases.R
#! /usr/bin/env Rscript
# Authors: Xindi Guo, Kristen Dang
# September 17, 2019

# setup --------------------------

library(glue)
library(sqldf)
library(synapser)
library(optparse)
library(plyr)
source("shared_fxns.R")

synLogin()

options(stringsAsFactors = FALSE)

# user input --------------------------

option_list <- list( 
  make_option(c("-i", "--input"), 
              type = "character",
              help="Synapse ID of the input file that has the BPC selected cases"),
  make_option(c("-o", "--output"), 
              type = "character",
              default = "syn20798271",
              help="Synapse ID of the BPC output folder. Default: syn20798271"),
  make_option(c("-p", "--phase"), 
              type = "character",
              help="BPC phase. i.e. pilot, phase 1, phase 1 additional"),
  make_option(c("-c", "--cohort"), 
              type = "character",
              help="BPC cohort. i.e. NSCLC, CRC, BrCa, and etc."),
  make_option(c("-s", "--site"), type = "character",
              help="BPC site. i.e. DFCI, MSK, UHN, VICC, and etc."),
  make_option(c("-r", "--release"), type = "character",
              help="Main GENIE clinical file release version name, e.g. 17.2-consortium.")
)
opt <- parse_args(OptionParser(option_list=option_list))

if (is.null(opt$input) || is.null(opt$phase) || is.null(opt$cohort) || is.null(opt$site)) {
  stop("Usage: Rscript export_bpc_selected_cases.R -h")
}

in_file <- opt$input
out_folder <- opt$output
phase <- opt$phase
cohort <- opt$cohort
site <- opt$site
release <- opt$release

# check user input -----------------

# TODO: how to get the options
# file_view_id <- "syn21557543"
# file_view_schema <- synGetColumns(file_view_id)$asList()

phase_option <- c("phase 1","phase 1 additional","phase 2")
cohort_option <- c("NSCLC","CRC","BrCa","PANC","Prostate","BLADDER", "MELANOMA", "RENAL", "OVARIAN", "ESOPHAGO")
site_option <- c("DFCI","MSK","UHN","VICC","UCSF","PROV","VHIO","JHU","WAKE")

# Check if phase is valid
phase_str <- paste0(phase_option, collapse = ", ")
if (!is.element(phase, phase_option)) {
  stop(glue("Error: {phase} is not a valid phase. Valid values: {phase_str}\nUsage: export_bpc_selected_cases.R -h"))
}

# Check if cohort is valid
cohort_str <- paste0(cohort_option, collapse = ", ")
if (!is.element(cohort, cohort_option)) {
  stop(glue("Error: {cohort} is not a valid cohort. Valid values: {cohort_str}\nUsage: export_bpc_selected_cases.R -h"))
}

# Check if site is valid
site_str <- paste0(site_option, collapse = ", ")
if (!is.element(site, site_option)) {
  stop(glue("Error: {site} is not a valid site. Valid values: {site_str}\nUsage: export_bpc_selected_cases.R -h"))
}

# setup ----------------------------
print("get clinical data")
# clinical data
# mapping tables
sex_mapping <- synTableQuery("SELECT * FROM syn7434222")$asDataFrame()
race_mapping <- synTableQuery("SELECT * FROM syn7434236")$asDataFrame()
ethnicity_mapping <- synTableQuery("SELECT * FROM syn7434242")$asDataFrame()
#sample_type_mapping <- synTableQuery("SELECT * FROM syn7434273")$asDataFrame()

# output setup
phase_no_space <- sub(" ","_",sub(" ","",phase))
output_entity_name <- glue("{site}_{cohort}_{phase_no_space}_genie_export.csv")
output_file_name <- glue("{site}_{cohort}_{phase_no_space}_genie_export_{Sys.Date()}.csv")

# download input file and get selected cases/samples
selected_info <- read.csv(synGet(in_file)$path)
selected_cases <- selected_info$PATIENT_ID
selected_samples <- unlist(strsplit(paste0(selected_info$SAMPLE_IDS,collapse=";"),";"))

# create the data file ----------------------------

# Create query for selected cases
temp <- toString(selected_samples)
temp <- sapply(strsplit(temp, '[, ]+'), function(x) toString(shQuote(x)))

# download clinical data
main_clinical <- get_main_genie_clinical_id(release = release)

# load and subset clinical data
clinical <- read.delim(synGet(main_clinical, downloadFile = TRUE, followLink = TRUE)$path, header = TRUE)
clinical <- sqldf(paste("SELECT * FROM clinical where SAMPLE_ID in (",temp,")",sep = ""))

# change the columns to lower case
colnames(clinical) <- tolower(colnames(clinical))
print("get all samples for selected patients")
# Get all samples for those patients
missing_patients <- selected_cases[!selected_cases %in% clinical$patient_id]
missing_samples <- selected_samples[!selected_samples %in% clinical$sample_id]
# Some patients/samples may not exist so filter those out
existing_patients = selected_cases[selected_cases %in% clinical$patient_id]
samples_per_patient <- clinical$sample_id[clinical$patient_id %in% selected_cases]

print("map data for each instrument")
# instrument - patient_characteristics
patient_output <- remap_patient_characteristics(clinical, existing_patients, ethnicity_mapping, race_mapping, sex_mapping)

print("recode")
# recode
# cannotReleaseHIPAA = NA
patient_output$birth_year[which(patient_output$birth_year == "cannotReleaseHIPAA")] <- NA
# -1 Not collected = 9 Unknown whether Spanish or not
patient_output$naaccr_ethnicity_code[which(patient_output$naaccr_ethnicity_code == -1)] <- 9
# -1 Not collected = 99 Unknown by patient
patient_output$naaccr_race_code_primary[which(patient_output$naaccr_race_code_primary == -1)] <- 99
# -1 Not collected = 88 No further race documented according to NAACCR
patient_output$naaccr_race_code_secondary[which(patient_output$naaccr_race_code_secondary == -1)] <- 88
patient_output$naaccr_race_code_tertiary[which(patient_output$naaccr_race_code_tertiary == -1)] <- 88

print("instrument cancer panel test")
# instrument - cancer_panel_test
sample_info_list <- lapply(samples_per_patient,function(x){
  sample_list = list()
  for(i in 1:length(x)){
    temp_df = data.frame("record_id" = clinical$patient_id[clinical$sample_id == x[i]])
    temp_df$redcap_repeat_instrument = "cancer_panel_test"
    temp_df$redcap_repeat_instance = i
    temp_df$redcap_data_access_group = clinical$center[clinical$sample_id == x[i]]
    
    temp_df$cpt_genie_sample_id = x[i]
    temp_df$cpt_oncotree_code = clinical$oncotree_code[clinical$sample_id == x[i]]
    temp_df$cpt_sample_type = clinical$sample_type_detailed[clinical$sample_id == x[i]]
    temp_df$cpt_seq_assay_id = clinical$seq_assay_id[clinical$sample_id == x[i]]
    temp_df$cpt_seq_date = clinical$seq_year[clinical$sample_id == x[i]]
    temp_df$age_at_seq_report = clinical$age_at_seq_report_days[clinical$sample_id == x[i]]
    sample_list[[i]] <- temp_df
  }
  combined_df = rbind.fill(sample_list)
  return(combined_df)
})

sample_info_df <- rbind.fill(sample_info_list)
patient_output <- rbind.fill(patient_output,sample_info_df)

print("validate output")
n_unique_patients_export = length(unique(sample_info_df$record_id))
n_unique_samples_export = length(unique(sample_info_df$cpt_genie_sample_id))
n_unique_selected_patients = length(unique(selected_cases))
n_unique_selected_samples = length(unique(selected_samples))
n_missing_patients = length(missing_patients)
n_missing_samples = length(missing_samples)
print("Patient(s) removed from the BPC export file:")
print(missing_patients)
print("Sample(s) removed from the BPC export file:")
print(missing_samples)
print(paste("N missing patients:", n_missing_patients))
print(paste("N missing samples:", n_missing_samples))
print(paste("N unique patients in export file", n_unique_patients_export))
print(paste("N unique samples in export file", n_unique_samples_export))
print(paste("N Unique patients in case selection", n_unique_selected_patients))
print(paste("N Unique samples in case selection", n_unique_selected_samples))

if (n_unique_samples_export != n_unique_selected_samples - n_missing_samples){
  stop("Number of unique samples in export file does not match number of samples in case selection")
}
if (n_unique_patients_export != n_unique_selected_patients - n_missing_patients){
  stop("Number of unique patients in export file does not match number of patients in case selection")
}
if (!all(patient_output$record_id %in% selected_cases)){
  stop("Some patients in export file are not in case selection")
}
# There is expected NA, because the export file is technically two csvs concatenated together
if (!all(unique(na.omit(patient_output$cpt_genie_sample_id)) %in% samples_per_patient)){
  stop("Some samples in export file are not in case selection")
}

print("output and upload")
# output and upload ----------------------------
write.csv(patient_output,file = output_file_name,quote = TRUE,row.names = FALSE,na = "")
act <- Activity(name = 'export main GENIE data', 
                description='Export selected BPC patient data from main GENIE clinical files',
                used = c(main_clinical, in_file),
                executed = 'https://github.com/Sage-Bionetworks/genie-bpc-pipeline/tree/develop/scripts/case_selection/export_bpc_selected_cases.R')
syn_file <- File(output_file_name, 
                 parent=out_folder,
                 name=output_entity_name,
                 annotations=list(phase=sub("phase ", "", phase),cohort=cohort,site=site))
syn_file <- synStore(syn_file)
synSetProvenance(syn_file,act)
file.remove(output_file_name)
